File: src/com/timepath/hl2/io/bsp/lump/Face.java
package com.timepath.hl2.io.bsp.lump;

import com.timepath.io.struct.StructField;
import java.util.logging.Logger;

public class Face {

    private static final Logger LOG = Logger.getLogger(Face.class.getName());

    private static final int MAXLIGHTMAPS = 4;

    /**
     * The plane number. unsigned
     */
    @StructField(index = 0) public short planenum;

    /**
     * faces opposite to the node's plane direction
     */
    @StructField(index = 1) public byte side;

    /**
     * 1 of on node, 0 if in leaf
     */
    @StructField(index = 2) public byte onNode;

    /**
     * index into surfedges
     */
    @StructField(index = 3) public int firstedge;

    /**
     * number of surfedges
     */
    @StructField(index = 4) public short numedges;

    /**
     * texture info
     */
    @StructField(index = 5) public short texinfo;

    /**
     * displacement info
     */
    @StructField(index = 6) public short dispinfo;

    /**
     * ?
     */
    @StructField(index = 7) public short surfaceFogVolumeID;

    /**
     * switchable lighting info
     */
    @StructField(index = 8) public byte[] styles = new byte[MAXLIGHTMAPS];

    /**
     * offset into lightmap lump
     */
    @StructField(index = 9) public int lightofs;

    /**
     * face area in units^2
     */
    @StructField(index = 10) public float area;

    /**
     * texture lighting info
     */
    @StructField(index = 11) public int[] m_LightmapTextureMinsInLuxels = new int[2];

    /**
     * texture lighting info
     */
    @StructField(index = 12) public int[] m_LightmapTextureSizeInLuxels = new int[2];

    /**
     * original face this was split from
     */
    @StructField(index = 13) public int origFace;

    /**
     * primitives. unsigned
     */
    @StructField(index = 14) public short m_NumPrims;

    /**
     * unsigned
     */
    @StructField(index = 15) public short firstPrimID;

    /**
     * lightmap smoothing group. unsigned
     */
    @StructField(index = 16) public int smoothingGroups;

}

File: src/main/kotlin/com/timepath/hl2/io/bsp/lump/Face.kt
package com.timepath.hl2.io.bsp.lump

import com.timepath.hl2.io.bsp.Lump
import com.timepath.hl2.io.bsp.LumpHandler
import com.timepath.io.OrderedInputStream
import com.timepath.io.struct.Struct
import com.timepath.io.struct.StructField

private val MAXLIGHTMAPS = 4
private val MAX_MAP_FACES = 65536

public class Face(
        /** The plane number. unsigned */
        @StructField(0) val planenum: Short = 0,
        /** faces opposite to the node's plane direction */
        @StructField(1) val side: Byte = 0,
        /** 1 of on node, 0 if in leaf */
        StructField(2) val onNode: Byte = 0,
        /** index into surfedges */
        StructField(3) val firstedge: Int = 0,
        /** number of surfedges */
        StructField(4) val numedges: Short = 0,
        /** texture info */
        StructField(5) val texinfo: Short = 0,
        /** displacement info */
        StructField(6) val dispinfo: Short = 0,
        /** ? */
        StructField(7) val surfaceFogVolumeID: Short = 0,
        /** switchable lighting info */
        StructField(8) val styles: ByteArray = ByteArray(MAXLIGHTMAPS),
        /** offset into lightmap lump */
        StructField(9) val lightofs: Int = 0,
        /** face area in units^2 */
        StructField(10) val area: Float = 0f,
        /** texture lighting info */
        StructField(11) val m_LightmapTextureMinsInLuxels: IntArray = IntArray(2),
        /** texture lighting info */
        StructField(12) val m_LightmapTextureSizeInLuxels: IntArray = IntArray(2),
        /** original face this was split from */
        StructField(13) val origFace: Int = 0,
        /** primitives. unsigned */
        StructField(14) val m_NumPrims: Short = 0,
        /** unsigned */
        StructField(15) val firstPrimID: Short = 0,
        /** lightmap smoothing group. unsigned */
        StructField(16) val smoothingGroups: Int = 0
) {
    class Handler : LumpHandler<List<Face>> {
        override fun invoke(l: Lump, ois: OrderedInputStream): List<Face> {
            return (0..(l.length / Struct.sizeof(Face())) - 1).map {
                ois.readStruct<Face>(Face())
            }
        }
    }
}
